mitiq/pt/pt.py

```diff
@@ -0,0 +1,163 @@
+# Copyright (C) Unitary Fund
+#
+# This source code is licensed under the GPL license (v3) found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Callable, List, Optional, Union, cast
+
+import random
+import cirq
+import numpy as np
+
+from mitiq import QPROGRAM, Executor, Observable, QuantumResult
+from mitiq.interface import noise_scaling_converter
+
+# P, Q, R, S from https://arxiv.org/pdf/2301.02690.pdf
+CNOT_twirling_gates = [
+    (cirq.I, cirq.I, cirq.I, cirq.I),
+    (cirq.I, cirq.X, cirq.I, cirq.X),
+    (cirq.I, cirq.Y, cirq.Z, cirq.Y),
+    (cirq.I, cirq.Z, cirq.Z, cirq.Z),
+    (cirq.Y, cirq.I, cirq.Y, cirq.X),
+    (cirq.Y, cirq.X, cirq.Y, cirq.I),
+    (cirq.Y, cirq.Y, cirq.X, cirq.Z),
+    (cirq.Y, cirq.Z, cirq.X, cirq.Y),
+    (cirq.X, cirq.I, cirq.X, cirq.X),
+    (cirq.X, cirq.X, cirq.X, cirq.I),
+    (cirq.X, cirq.Y, cirq.Y, cirq.Z),
+    (cirq.X, cirq.Z, cirq.Y, cirq.Y),
+    (cirq.Z, cirq.I, cirq.Z, cirq.I),
+    (cirq.Z, cirq.X, cirq.Z, cirq.X),
+    (cirq.Z, cirq.Y, cirq.I, cirq.Y),
+    (cirq.Z, cirq.Z, cirq.I, cirq.Z),
+]
+CZ_twirling_gates = [
+    (cirq.I, cirq.I, cirq.I, cirq.I),
+    (cirq.I, cirq.X, cirq.Z, cirq.X),
+    (cirq.I, cirq.Y, cirq.Z, cirq.Y),
+    (cirq.I, cirq.Z, cirq.I, cirq.Z),
+    (cirq.X, cirq.I, cirq.X, cirq.Z),
+    (cirq.X, cirq.X, cirq.Y, cirq.Y),
+    (cirq.X, cirq.Y, cirq.Y, cirq.X),
+    (cirq.X, cirq.Z, cirq.X, cirq.I),
+    (cirq.Y, cirq.I, cirq.Y, cirq.Z),
+    (cirq.Y, cirq.X, cirq.X, cirq.Y),
+    (cirq.Y, cirq.Y, cirq.X, cirq.X),
+    (cirq.Y, cirq.Z, cirq.Y, cirq.I),
+    (cirq.Z, cirq.I, cirq.Z, cirq.I),
+    (cirq.Z, cirq.X, cirq.I, cirq.X),
+    (cirq.Z, cirq.Y, cirq.I, cirq.Y),
+    (cirq.Z, cirq.Z, cirq.Z, cirq.Z),
+]
+
+
+def execute_with_pauli_twirling(
+    circuit: QPROGRAM,
+    executor: Union[Executor, Callable[[QPROGRAM], QuantumResult]],
+    observable: Optional[Observable] = None,
+    *,
+    num_circuits: int = 10,
+) -> float:
+    """Estimates the expecation value by averaging pauli twirled circuits.
+
+    Args:
+        circuit: The input circuit to execute with twirling.
+        executor: A Mitiq executor that executes a circuit and returns the
+            unmitigated ``QuantumResult`` (e.g. an expectation value).
+        observable: Observable to compute the expectation value of. If
+            ``None``, the ``executor`` must return an expectation value
+            (float). Otherwise, the ``QuantumResult`` returned by ``executor``
+            is used to compute the expectation of the observable.
+        num_circuits: Number of circuits to be randomly twirled, and averaged.
+
+    Returns:
+        The expectation value estimated with Pauli twirling.
+    """
+    executor = (
+        executor if isinstance(executor, Executor) else Executor(executor)
+    )
+    twirled_circuits = twirl_CNOT_gates(circuit, num_circuits)
```

Comment:
Why is only `twirl_CNOT_gates` used in `execute_with_pauli_twirling` and not `twirl_CZ_gates`?

Reply:
Good catch. I added CZ twirling in https://github.com/unitaryfund/mitiq/pull/1802/commits/4d1e9b1cfaadfd19d48026f3f16877b37a2fa3a9.